### ISSUE TITLE
Division by num_positive_anchors to get average cls_loss seems wrong?

### ISSUE THREAD

**Issue by gunshi** [yhenon/pytorch-retinanet#98]
Hey, thanks for your code!
I have a question about the line here: 
https://github.com/yhenon/pytorch-retinanet/blob/1135e18b835481b18fd0d4e1613c87afc2bc7d46/losses.py#L92

Is that correct? Shouldn't we be dividing by the total number of elements in 'targets' not equal to -1 instead? Since num_positive_anchors is just the elements which had binary class =1, but we also get cls_loss from elements with binary class =0.
Would greatly appreciate clarification on this.
Thanks
Gunshi

**Comment by CanshangD** [comment]
I think that because only positive sample loss is calculated， it is reasonable to divide loss by num_positive_anchors.

**Comment by zjysteven** [comment]
@gunshi I had the same question. However, when I tried to divide the cls_loss by the total number of both background and positive anchors as you said, the cls_loss turned out to be extremely small (~0.0010) even from the start of the training. This is not surprising as the number of background anchors is so large. In this case, such a small loss value will make the model learn nothing, as my trained model output classification probabilities all around 0. When changing back to dividing with the number of positive anchors only, I got reasonable outputs. I guess that's a possible reason for doing this. I also asked keras-retinanet's authors for clarification on this, you can find it here https://github.com/fizyr/keras-retinanet/issues/1254.

@CanshangD Actually the background sample loss is also calculated (this is also why focal loss is proposed for), it's just that their targets are all zeros instead of one-hot labels for positive samples.

**Comment by yhenon** [comment]
The paper says:
>The  total  focal  loss  of  an  image  is  computed  as  the  sum of the focal loss over all ∼100k anchors, normalized by the number of anchors assigned to a ground-truth box

Which is what I implemented. I agree it is a little surprising though.
